drivers: modem: ublox-sara-r4: autodetection of modem and APN

hwilmers opened the pull request:
Implemented:
- automatic detection of modem type (Sara R410 or Sara U201)
- automatic setting of APN, depending on inserted SIM card

Automatic detection of the modem type:

After hardware initialisation of the modem, the type of modem
is detected. After this, further initialisation and functionality
is chosen depending on the detected modem type.

To enable this functionality, set following bool in Kconfig:
   MODEM_UBLOX_SARA_AUTODETECT_VARIANT

Automatic setting of APN:

During communication initialisation, the IMSI of the inserted SIM
card is evaluated to determine the APN. This is done by comparing
the first 5 characters of the IMSI to a list of known providers.
The list can be given in Kconfig.

To enable this functionality, set following bool in Kconfig:
   MODEM_UBLOX_SARA_AUTODETECT_APN

To set a list of providers, set following string:
   MODEM_UBLOX_SARA_AUTODETECT_APN_PROFILES

If the provider can not be found in the list, the APN given in
following entry is used as a fallback:
   MODEM_UBLOX_SARA_R4_APN

Signed-off-by: Hans Wilmers <hans@wilmers.no>

hwilmers commented:
Caveat: When autodetection is chosen, it is presumed that initialisation of the hardware can be identical for R410 and U201. On my hardware this is the case, but probably it is not true generally.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1766,4 +1766,3 @@ NET_DEVICE_OFFLOAD_INIT(modem_sara, CONFIG_MODEM_UBLOX_SARA_R4_NAME,
 			modem_init, &mdata, NULL,
 			CONFIG_MODEM_UBLOX_SARA_R4_INIT_PRIORITY, &api_funcs,
 			MDM_MAX_DATA_LENGTH);
-
```
Just squash the "drivers: modem: ublox-sara-r4: removed blank line" commit to previous one.

hwilmers replied:
@jukkar Sorry - tried to do that, but it only added more commits. Should I clear this up some way or another?

jukkar replied:
Just squash the commits together, and force push to this branch. Do not close this PR and resubmit, it is just pointless.

jukkar replied:
Are you trying to do this from github web ui? That (squashing, force pushing) is probably not going to work from web ui, one needs to use git command line for that.

hwilmers replied:
got it - thanks for the comments.

jukkar reviewed (commented):
We have also the generic GSM modem support and something like this could be implemented there too (if using ublox-sara modem in non-offloading case). In order not to duplicate too much code, I just wonder if we could have some generic way of adding this kind of support.

jukkar commented on drivers/modem/Kconfig.ublox-sara-r4:
```diff
@@ -60,6 +65,19 @@ config MODEM_UBLOX_SARA_R4_APN
+	string "list of profiles to search when autodetecting APN"
+	default "hologram=23450, wm=20601 29505 29509 23450 53703 90143"
+	help
+		Set a comma separated list of profiles, each containing of:
```
Please indent the help lines like the other help blocks, so with two spaces

hwilmers replied:
done

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -190,6 +212,86 @@ static int modem_atoi(const char *s, const int err_value,
 
+#if defined(CONFIG_MODEM_UBLOX_SARA_AUTODETECT_APN)
+
+const char *modem_sim_profiles =
```
perhaps static here?

hwilmers replied:
I need to be able to change this on application startup. The least intrusive way I see is to use global scope - but if we should do it in another way, please comment.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -190,6 +212,86 @@ static int modem_atoi(const char *s, const int err_value,
+			s--;
+		}
+
+		char *eos = s+1;
```
Introduce the variable at the beginning of the block, not middle.

hwilmers replied:
done

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -687,6 +819,42 @@ static int pin_init(void)
+						 &mdata.sem_response,
+						 MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("AT+C[E]SQ ret:%d", ret);
```
If we have an error, should we bail out here?

hwilmers replied:
we don't bail out here, because we want to restart a new RSSI query, regardless of teh outcome of this one. This is also how it is done in the original code.

hwilmers replied:
not really - we want to start the next query afterwards, regardless the outcome of this one.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -799,6 +994,25 @@ static void modem_reset(void)
 
+#if defined(CONFIG_MODEM_UBLOX_SARA_AUTODETECT_APN)
+	/* autodetect APN from IMSI */
+	char cmd[64];
```
Instead of this hard coded value, I think something like below would be better (user does not have to count the chars itself as the compiler is doing that for you). The APN name just needs to be the longest name possible.
```
char cmd[sizeof("AT+CGDCONT=1,'IP','012345678901234567890123456789012'")];
```

hwilmers replied:
good point! I changed it to use the correct (max) length now.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -876,11 +1090,37 @@ static void modem_reset(void)
+
+#if defined(CONFIG_MODEM_UBLOX_SARA_AUTODETECT_APN)
+		/* setup PDP context definition */
+		char cmd[64];
```
Similar thing here

hwilmers replied:
Thank you for your input. I will look at this within a few days. It it ok for the workflow if I add commits, or should I squash them again, replacing the original commit?

jukkar replied:
As a general rule, if you are fixing things in your own commits, then squash. If you are introducing new logically separate stuff, then do new commit. In this case, just squash the commits.

jukkar commented:
ping @mike-scott

jukkar commented:
@mike-scott does this look ok to you?

mike-scott commented:
@jukkar I'll play test today.  Thanks for the reminder.

mike-scott commented:
@hwilmers After reviewing this, I think this should be 2 commits:
1. automatic detection of modem type (Sara R410 or Sara U201) changes
2. and then automatic setting of APN, depending on inserted SIM card

I also am questioning the complexity of the #ifdefs in this driver. If we can come up with a creative solution for the different pin init between U2 and R4, then we could do away with the manual selection of the modem type entirely.

But let's get the 2 changes into their own commits first.

hwilmers committed:
drivers: modem: ublox-sara-r4: automatic detection of modem type

The modem type (Sara R410 or Sara U201) is detected automatically after
hardware initialisation of the modem. Further initialisation and
functionality is then chosen depending on the detected modem type.

To enable this functionality, set following bool in Kconfig:
MODEM_UBLOX_SARA_AUTODETECT_VARIANT

Signed-off-by: Hans Wilmers <hans@wilmers.no>

hwilmers committed:
drivers: modem: ublox-sara-r4: automatic setting of APN

During communication initialisation, the IMSI of the inserted SIM
card is evaluated to determine the APN. This is done by comparing
the first 5 characters of the IMSI to a list of known providers.
The list can be given in Kconfig.

To enable this functionality, set following bool in Kconfig:
MODEM_UBLOX_SARA_AUTODETECT_APN

To set a list of providers, set following string:
MODEM_UBLOX_SARA_AUTODETECT_APN_PROFILES

If the provider can not be found in the list, the APN given in
following entry is used as a fallback:
MODEM_UBLOX_SARA_R4_APN

Signed-off-by: Hans Wilmers <hans@wilmers.no>

hwilmers commented:
@mike-scott Changes are now divided in two commits. Would you have any suggestions about how to reduce complexity of the #ifdefs? I am reluctant to introduce dead code, but if it would be ok to introduce some, complexity could be reduced.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -687,6 +711,42 @@ static int pin_init(void)
+		  MODEM_CMD("+CSQ: ", on_cmd_atcmdinfo_rssi_csq, 2U, ","),
+		  MODEM_CMD("+CESQ: ", on_cmd_atcmdinfo_rssi_cesq, 6U, ","),
+	};
+	static char *send_cmd_u2 = "AT+CSQ";
```
Shouldn't these two vars be const instead of static?

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -687,6 +711,42 @@ static int pin_init(void)
+	int ret;
+
+	/* choose cmd according to variant */
+	char *send_cmd = send_cmd_r4;
```
also this could be const

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -687,6 +711,42 @@ static int pin_init(void)
+
+	/* query modem RSSI */
+	ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+						 cmds, ARRAY_SIZE(cmds),
```
indentation

jukkar commented on drivers/modem/ublox-sara-r4.c:
Indentation. Is your tab size correct, it should be 8

jukkar commented on drivers/modem/ublox-sara-r4.c:
ditto

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -872,6 +999,25 @@ static void modem_reset(void)
+
+	/* setup PDP context definition */
+	ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+						 NULL, 0,
```
ditto

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -872,6 +999,25 @@ static void modem_reset(void)
+						 MDM_REGISTRATION_TIMEOUT);
+
+	ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+						 NULL, 0,
```
here too

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -951,6 +1097,18 @@ static void modem_reset(void)
+
+		snprintf(cmd, sizeof(cmd), "AT+UPSD=0,1,\"%s\"", mdata.mdm_apn);
+		ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+					 NULL, 0,
```
and here

mike-scott commented:
@hwilmers Thanks for splitting this up.  I was thinking about the exact same thing with the `#ifdef` complexity.  If we could consolidate the init process for both R4 and U2, then we would remove the 3 `#ifdef`s which are really the only sticking point to removing the variant `#ifdef`s entirely.  It would increase the code size of the driver by an unknown amount (not a huge amount) -- but would also provide good code coverage.

mike-scott commented:
@hwilmers Let me look at the Application notes power on/off sequence for both the R4xx and U2xx and see if we can conslidate.

hwilmers commented:
@mike-scott Is there a chance to get this in before the merge window closes?

jukkar commented:
I think we could try to get this to 2.3, @mike-scott please ack if ok.

carlescufi commented:
@jukkar I'd need at least one approval for this to go in. I will add @rlubos as well in case he can take a look.

mike-scott commented:
I'll take another look

mike-scott commented:
@hwilmers @jukkar @carlescufi I think this PR still needs a bit of discussion:

My current issues are:
1. The R4 vs. U2 differences in `pin_init()` are not addressed.  I think in it's current form, this PR would probably break R4.
2. The second patch might be mergeable as a stand-alone patch, but it only applies to U2 which uses the AT+UPSD command.  R4 relies on AT+CGDCONT to set APN.

hwilmers commented:
>     1. The R4 vs. U2 differences in `pin_init()` are not addressed.  I think in it's current form, this PR would probably break R4.
> 
Yes, I see that there is a problem. But it is the other way around, since with this patch (but only with MODEM_UBLOX_SARA_AUTODETECT_VARIANT set), for both modems the pin_init() is done as though they were R4. I have no better solution for this, and because on my hardware pin_init() is merely ignored, I can not test for this case.

>     2. The second patch might be mergeable as a stand-alone patch, but it only applies to U2 which uses the AT+UPSD command.  R4 relies on AT+CGDCONT to set APN.

If R4 is detected, AT+CGDCONT is used to set the APN. I apologise for the complicated #ifdef logic that makes this less obvious than it could be.

mike-scott commented:
@hwilmers 
> > ```
> > 1. The R4 vs. U2 differences in `pin_init()` are not addressed.  I think in it's current form, this PR would probably break R4.
> > ```
> 
> Yes, I see that there is a problem. But it is the other way around, since with this patch (but only with MODEM_UBLOX_SARA_AUTODETECT_VARIANT set), for both modems the pin_init() is done as though they were R4. I have no better solution for this, and because on my hardware pin_init() is merely ignored, I can not test for this case.

I would also hate to break U2 users.  I don't think we can merge the first patch until we figure out the pin_init changes. :/

> 
> > ```
> > 2. The second patch might be mergeable as a stand-alone patch, but it only applies to U2 which uses the AT+UPSD command.  R4 relies on AT+CGDCONT to set APN.
> > ```
> 
> If R4 is detected, AT+CGDCONT is used to set the APN. I apologise for the complicated #ifdef logic that makes this less obvious than it could be.

Oh you are correct.  The mdm_apn value is also used for AT+CGDCONT (somehow I missed this initially).  So the 2nd patch can probably be merged.  Do you want to submit that as a separate patch in a new PR or perhaps remove the first patch from here and re-post that patch as a new PR so it can continue to develop?

mike-scott commented:
@hwilmers The first patch is going to need the following:
1. A U2 user to confirm whatever pin_init changes are working for them
2. I can validate the R4 pin_init changes on my end

hwilmers commented:
@mike-scott About the first patch: I may be missing something here, but there are no changes in pin_init() at all. If you choose to set MODEM_UBLOX_SARA_U2 or MODEM_UBLOX_SARA_R4, it will behave exactly as before.

But I don't want to force this onto anyone. If you think it is better, I can remove this patch from the PR.

mike-scott commented:
> @mike-scott About the first patch: I may be missing something here, but there are no changes in pin_init() at all. If you choose to set MODEM_UBLOX_SARA_U2 or MODEM_UBLOX_SARA_R4, pin_init() will behave exactly as before.
> 
> But I don't want to force this onto anyone. If you think it is better, I can remove this patch from the PR.

@hwilmers I should have worded my original response this way:

I wouldn't want to merge the first patch and have users try it with unexpected results.  I would rather merge it when it's ready for all users.

You are correct in that by itself it doesn't actually break anyone.  As long as they don't swap to using `MODEM_UBLOX_SARA_AUTODETECT_VARIANT`.

I'm curious how you are able to use the first patch as it is now.  You mention that for your hardware pin_init() is ignored?  What HW is that?  If there is a valid user of this patch, then I would be more inclined to merge it now and then work to cleanup `pin_init()`.

hwilmers commented:
> I'm curious how you are able to use the first patch as it is now. You mention that for your hardware pin_init() is ignored? What HW is that? If there is a valid user of this patch, then I would be more inclined to merge it now and then work to cleanup `pin_init()`.
I am using cellular XBee modules made by Digi. The hardware init of the modem engines is done by by some middle layer made by Digi, but after hardware init they allow to communicate directly to the modem engines (in this case U201 or R410). This is why pin_init() works identical for both types - it is merely ignored.

mike-scott commented:
@hwilmers if you say that this PR makes the cellular XBee modules made by Digi usable with Zephyr.  Then I'm ok with +1 for merge in 2.3.

mike-scott reviewed (approved):
Let's add this for 2.3 to enable support for cellular XBee modules made by Digi.
